content/part4.md

```diff
@@ -0,0 +1,199 @@
+Even though `workflowID` is unique in this case, you still need to provide `runID` in order to Query or Signal the cart Workflow.
+The next endpoint is `GET /cart/{workflowID}/{runID}`, which returns the current state of the cart with the given `WorkflowID` and `runID`.
+Below is the `GetCartHandler()` function, which gets the `workflowID` and `runID` from the URL and executes a Query for the current state of the cart.
+
+```go
+func GetCartHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	response, err := temporal.QueryWorkflow(context.Background(), vars["workflowID"], vars["runID"], "getCart")
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+	var res interface{}
+	if err := response.Get(&res); err != nil {
+		WriteError(w, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(res)
+}
+```
+
+## PUT Requests and Signals
+
+HTTP PUT requests usually correspond to Temporal Signals.
+That means, in addition to the `workflowID` and `runID`, you need to send Signal arguments.
+Remember that `shared.go` contains an [`AddToCartSignal` struct](https://github.com/temporalio/temporal-ecommerce/blob/5c4e0142e3571398d972c80b3fa7cdbe7a5db42b/shared.go#L64-L67) that is what the [cart Workflow's Signal handler expects](https://github.com/temporalio/temporal-ecommerce/blob/main/workflow.go#L52-L71):
+
+```go
+type AddToCartSignal struct {
+	Route string
+	Item  CartItem
+}
+```
+
+The `PUT /cart/{workflowID}/{runID}/add` handler needs to convert the HTTP request body into an `AddToCartSignal` as shown below.
+
+```go
+func AddToCartHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	var item app.CartItem
+	err := json.NewDecoder(r.Body).Decode(&item)
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+
+	update := app.AddToCartSignal{Route: app.RouteTypes.ADD_TO_CART, Item: item}
+
+	err = temporal.SignalWorkflow(context.Background(), vars["workflowID"], vars["runID"], app.SignalChannelName, update)
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+
+  // Signals don't return a response, so assume that if the Signal went through
+  // then the request succeeded.
+	w.WriteHeader(http.StatusOK)
+	res := make(map[string]interface{})
+	res["ok"] = 1
+	json.NewEncoder(w).Encode(res)
+}
+```
+
+The `PUT /cart/{workflowID}/{runID}/remove` and `PUT /cart/{workflowID}/{runID}/email` handlers are almost identical, except they send `RemoveFromCartSignal` and `UpdateEmailSignal`, not `AddToCartSignal`.
+
+```go
+func UpdateEmailHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+
+	var body UpdateEmailRequest
+	err := json.NewDecoder(r.Body).Decode(&body)
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+
+	updateEmail := app.UpdateEmailSignal{Route: app.RouteTypes.UPDATE_EMAIL, Email: body.Email}
+
+	err = temporal.SignalWorkflow(context.Background(), vars["workflowID"], vars["runID"], app.SignalChannelName, updateEmail)
```

Comment:
Temporal doesn't really support the same channel for multiple record types. Unless I'm missing something but SignalChannelName is used for both `UpdateEmailSignal` and `AddToCartSignal`.